### PR TITLE
Combine statuses Failing, Failed, Error, Warning, FailedToGet and MulipleJobsRunning into one "Problem"

### PR DESCRIPTION
**Pull request by gskrobisz**
Changes:
1. Statuses Failing, Failed, Error, Warning, FailedToGet and MultipleJobsRunning are replaced by one dynamic status `ProblemStateStatus`.
2. Removed `SimpleStateStatus.Undefined`.
3. For each of "Problem" statuses tooltip and description have the same value (to remove unnecessary "UI message clutter" - each status had different tooltip and description, almost the same words but different arrangement)

Other changes:
1. Parameter `delegate` in `OverridingProcessStateDefinitionManager` now has no default value. For clarity it should be provided explicitly.

To discuss / hint needed:
1. Better name and location for `ProblemStateStatus`.
2. Better location for MultipleJobsRunning - it was part of Flink&K8 custom statuses, now it is in SimpleStateStatus.ProblemStateStatus which seems to be wrong. Maybe it is better to move `ProblemStateStatus` outside of `SimpleStateStatus`.
3. Remove unnecessary icons (keep error.svg instead of failed.svg? keep both "just in case"?)
4. Provide descriptive "how to handle problem and what to do next" infos when status "Problem" occurs. Current descriptions/tooltips do not provide helpful information what to do with the problem, e.g. "Scenario has been canceled but still is running" or "Scenario deployed in version ... (by ...), should be running".
5. Should healthcheck deployed-not-running indicate occurences of "Problem" scenarios?

Issues moved to another PR
6. StateStatus.isDuringDeploy can be replaced with pattern match with DuringDeployStateStatus, we do not need this at StateStatus level.
7. ProcessState.isDeployed - why during deploy is recognized as deployed?

### CODE REVIEW

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/OverridingProcessStateDefinitionManager.scala [hunk 1]
```diff
@@ -20,13 +19,12 @@ import java.net.URI
   *
   * @param customStateDefinitions Set of definitions that extends or overwrites delegate definitions
   */
-class OverridingProcessStateDefinitionManager(statusActionsPF: PartialFunction[StateStatus, List[ProcessActionType]] = PartialFunction.empty,
-                                              mapActionToStatusPF: PartialFunction[Option[ProcessActionType], StateStatus] = PartialFunction.empty,
+class OverridingProcessStateDefinitionManager(delegate: ProcessStateDefinitionManager,
+                                              statusActionsPF: PartialFunction[StateStatus, List[ProcessActionType]] = PartialFunction.empty,
```
Comment (arkadius) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/OverridingProcessStateDefinitionManager.scala:23]: Uhh I see that I left this - thanks for cleaning after my change!

File: docs/MigrationGuide.md [hunk 2]
```diff
@@ -27,6 +27,13 @@ To see the biggest differences please consult the [changelog](Changelog.md).
   which provide the same interface as the previous one, with only method name changed.
   Especially, when you use 'PeriodicDeploymentManagerProvider', `delegate` should already return `DeploymentManager` wrapped by caching mechanism.
 * [#4131](https://github.com/TouK/nussknacker/pull/4131) `Parameter.defaultValue` now holds `Option[Expression]` instead of `Option[String]`. You have to wrap a `String` with `Expression.spel()`
+* [#4132](https://github.com/TouK/nussknacker/pull/4132) 
+  * Within the base set of statuses used in Embedded, Flink, K8 and Periodic mode (`SimpleStateStatus`), statuses `Failing`, `Failed`, `Error`, `Warning`, `FailedToGet` and `MulipleJobsRunning`
+    are replaced by one `ProblemStateStatus` which is parametrized by specific message. `ProblemStateStatus` provides several builder methods, one for each corresponding removed state.
+    Those builders allow to preserve the exact moments when each state appears in the scenario lifecycle. 
+  * Displayed tooltip and description of `ProblemStateStatus` have the same value. 
+  * Removed `SimpleStateStatus.Undefined`
```
Comment (arkadius) [inline docs/MigrationGuide.md:35]: I can't see this status. Where it was?

File: docs/MigrationGuide.md [hunk 3]
```diff
@@ -27,6 +27,13 @@ To see the biggest differences please consult the [changelog](Changelog.md).
   which provide the same interface as the previous one, with only method name changed.
   Especially, when you use 'PeriodicDeploymentManagerProvider', `delegate` should already return `DeploymentManager` wrapped by caching mechanism.
 * [#4131](https://github.com/TouK/nussknacker/pull/4131) `Parameter.defaultValue` now holds `Option[Expression]` instead of `Option[String]`. You have to wrap a `String` with `Expression.spel()`
+* [#4132](https://github.com/TouK/nussknacker/pull/4132)
```
Comment (arkadius) [inline docs/MigrationGuide.md:30]: WDYT about somehow merge'ing this entry with #4122? So developer will see all changes in statuses in one place. Or at least move it close to it?

File: docs/MigrationGuide.md (hunk 2 above)
Comment (gskrobisz) [reply docs/MigrationGuide.md]: Fixed (Undefined -> Unknown)

**Review by arkadius — APPROVED** [review]
LGTM

**Review by gadomsky — APPROVED** [review]

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala [hunk 4]
```diff
@@ -2,54 +2,99 @@ package pl.touk.nussknacker.engine.api.deployment.simple
...
+
+    def processWithoutAction: ProblemStateStatus =
+      ProblemStateStatus(processWithoutActionMessage)
+
+    def multipleJobsRunning: ProblemStateStatus =
+      ProblemStateStatus(errorMultipleJobsMessage, isRedeployable = false)
+
+    // Problem descriptions
+
+    private def errorFailToGetStatus: String =
```
Comment (lciolecki) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala:48]: These belove methods' strings can be moved to the above methods?

File: docs/Changelog.md [hunk 5]
```diff
@@ -33,6 +33,8 @@
   On the other hand, the scenario editor presented state based on engine (e.g. Flink) state.
   After the change, in all places the state is based on engine (e.g. Flink) state - in some places like scenario list, it is cached.
 * [#4131](https://github.com/TouK/nussknacker/pull/4131) Support for components using other languages than SpEL, added basic support for SpEL in template mode
+* [#4132](https://github.com/TouK/nussknacker/pull/4132) Combine statuses Failing, Failed, Error, Warning, FailedToGet and MulipleJobsRunning into one status that represents a "Problem".
```
Comment (lciolecki) [inline docs/Changelog.md:36]: What about Unknown?

File: engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkStateStatus.scala [hunk 6]
```diff
@@ -1,47 +1,24 @@
...
 import pl.touk.nussknacker.engine.api.deployment.simple.SimpleStateStatus
 
-import java.net.URI
-
 object FlinkStateStatus {
-  val FailedToGet: StateStatus = SimpleStateStatus.FailedToGet
-  val Unknown: StateStatus = SimpleStateStatus.Unknown
+
+  // Flink statuses are based on SimpleStateStatus definitions
+  // TODO: remove those states here and use SimpleStateStatus
```
Comment (lciolecki) [inline engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkStateStatus.scala:10]: Why do you don't do this now? :)

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 7]
```diff
@@ -31,9 +32,7 @@ class K8sDeploymentStatusMapper(definitionManager: ProcessStateDefinitionManager
     deployments match {
       case Nil => None
       case one :: Nil => Some(status(one, pods))
-      case duplicates =>
-        val errors = List(s"Expected one deployment, instead: ${duplicates.map(_.metadata.name).mkString(", ")}")
-        Some(definitionManager.processState(K8sStateStatus.MultipleJobsRunning, errors = errors))
+      case duplicates => Some(K8sProcessStateDefinitionManager.errorMultipleJobsRunning(duplicates))
```
Comment (lciolecki) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala:35]: eg. on multiple you should be able to cancel scenario ;)

**Review by lciolecki — APPROVED** [review]
Nice, I left some mirror comments.

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 7 above)
Comment (gskrobisz) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala:35]: Yes. Action cancel is allowed in `SimpleStateStatus.statusActionsPF`

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala [hunk 8]
```diff
@@ -2,54 +2,99 @@ package pl.touk.nussknacker.engine.api.deployment.simple
...
+import pl.touk.nussknacker.engine.api.process.VersionId
 
 import java.net.URI
 
 object SimpleStateStatus {
-  val FailedToGet: StateStatus = NotEstablishedStateStatus("FAILED_TO_GET")
-  val Unknown: StateStatus = NotEstablishedStateStatus("UNKNOWN")
+
+  // Represents general problem.
+  case class ProblemStateStatus(description: String, isRedeployable: Boolean = true) extends CustomStateStatus(ProblemStateStatus.name) {
```
[the remark this answers is not in the document]
Comment (gskrobisz) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala:13]: See configuration of `SimpleStateStatus.statusActionsPF` where is defined mapping status to allowed actions. Flag `isRedeployable` is used to distinguish recoverable (see automatic `restartStrategy` that works only when Deploy action is allowed) from non-recoverable problems (automatic restart is not possible).

Before this change it was defined as custom "mapStateToAction" in FlinkStateStatus or K8StateStatus, where:
```
case FlinkStateStatus.MultipleJobsRunning => List(ProcessActionType.Cancel)
```
Now that MultipleJobsRunning is replaced by ProblemStateStatus (and we want to have only one "problem") with specific problem configuration.

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala (hunk 4 above)
Comment (gskrobisz) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/simple/SimpleStateStatus.scala]: Fixed.

**Pull request merged**